File: app/src/main/java/com/riders/thelab/ui/kat/KatActivity.java
package com.riders.thelab.ui.kat;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.riders.thelab.core.utils.LabDeviceManager;
import com.riders.thelab.core.utils.UIManager;
import com.riders.thelab.data.remote.dto.kat.Kat;
import com.riders.thelab.databinding.ActivityKatBinding;
import com.riders.thelab.ui.base.SimpleActivity;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.Random;

import timber.log.Timber;

public class KatActivity extends SimpleActivity implements TextWatcher {

    private ActivityKatBinding viewBinding;

    private List<Kat> fetchedKatMessageList;
    private List<Kat> currentMessageList;
    private KatMessagesAdapter mAdapter;

    private FirebaseDatabase database;
    private DatabaseReference messageRef;

    private Kat katModel;

    public static String SENDER_ID =
            LabDeviceManager.getBrand() + ", " + LabDeviceManager.getModel();

    @Override
    protected void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewBinding = ActivityKatBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());

        viewBinding.btnSendMessage.setEnabled(false);

        initMessageRecyclerView();

        // Write a message to the database
        database = FirebaseDatabase.getInstance();
        messageRef = database.getReference("messages");

        Timber.d("setupView()");
        Query messagesQuery = messageRef.getRef();

        // Build object
        initKatObject();


        // Read from the database
        messageRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NotNull DataSnapshot dataSnapshot) {
                Timber.d("ValueEventListener - onDataChange()");

                // This method is called once with the initial value and again
                // whenever data at this location is updated.
                GenericTypeIndicator<HashMap<String, Object>> genericTypeIndicator =
                        new GenericTypeIndicator<HashMap<String, Object>>() {
                        };
                HashMap<String, Object> value = dataSnapshot.getValue(genericTypeIndicator);

                // Get keys and values
                if (null != value) {

                    fetchedKatMessageList = new ArrayList<>();
                    fetchedKatMessageList = Kat.buildKatMessagesList(value);
                    updateUI();
                } else {
                    removeAll();
                }
            }

            @Override
            public void onCancelled(@NotNull DatabaseError error) {
                // Failed to read value
                Timber.e(error.toException(), "Failed to read value.");
            }
        });

        viewBinding.etMessage.addTextChangedListener(this);


        // Handle send message button event
        viewBinding.btnSendMessage.setOnClickListener(
                view -> {
                    String messageContent =
                            !Objects.requireNonNull(viewBinding.etMessage.getText()).toString().trim().isEmpty()
                                    ? viewBinding.etMessage.getText().toString().trim()
                                    : "";

                    clearEditTextField();

                    UIManager.hideKeyboard(KatActivity.this, findViewById(android.R.id.content));

                    // Add complement values to kat object
                    katModel.setMessage(messageContent);
                    katModel.setTimestamp(System.currentTimeMillis());

                    String key = messageRef.push().getKey();
                    katModel.setMessageId(key);

                    messageRef
                            .child(key)
                            .setValue(katModel);
                });
    }

    private void initMessageRecyclerView() {
        Timber.d("initMessageRecyclerView()");

        currentMessageList = new ArrayList<>();

        mAdapter = new KatMessagesAdapter(this, currentMessageList);

        LinearLayoutManager linearLayoutManager =
                new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        viewBinding.rvMessages.setLayoutManager(linearLayoutManager);
        viewBinding.rvMessages.setAdapter(mAdapter);
    }

    private void initKatObject() {
        Timber.d("initKatObject()");
        katModel = new Kat();
        katModel.setSenderId(SENDER_ID);
        katModel.setChatId(String.valueOf(new Random().nextInt(1000)));
    }

    private void clearEditTextField() {
        Timber.d("clearEditTextField()");
        viewBinding.etMessage.setText("");
    }

    private void updateUI() {
        Timber.d("updateUI()");

        Timber.d("isFromEventListener");
        // If it's first time, currentList is empty then populate list with fetched item
        if (currentMessageList.isEmpty()) {
            Timber.d("is first time");
            currentMessageList = fetchedKatMessageList;
            mAdapter.populateAllItems(currentMessageList);
        } else {
            Timber.e("NOT first time");
            Kat kat = fetchedKatMessageList.get(fetchedKatMessageList.size() - 1);
            Timber.e("last item : %s", kat.toString());
            mAdapter.populateItem(kat);
            smoothScrollToLastItem();
        }

    }

    private void smoothScrollToLastItem() {
        runOnUiThread(() -> viewBinding.rvMessages.smoothScrollToPosition(
                Objects.requireNonNull(viewBinding.rvMessages.getAdapter()).getItemCount() - 1));
    }

    private List<Kat> compareListAndGetAddedItem(List<Kat> currentList, List<Kat> fetchedList) {
        Timber.d("compareListAndGetAddedItem()");
        List<Kat> differences = new ArrayList<>(fetchedList);
        differences.removeAll(currentList);

        for (Kat katElement : differences) {
            Timber.e("Differences : %s", katElement.toString());
        }

        return differences;
    }

    private void removeAll() {
        mAdapter.removeAll();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        database = null;
        messageRef = null;
        fetchedKatMessageList = null;
        currentMessageList = null;
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        viewBinding.btnSendMessage.setEnabled(s.toString().trim().length() > 0);
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        viewBinding.btnSendMessage.setEnabled(s.toString().trim().length() > 0);
    }

    @Override
    public void afterTextChanged(Editable s) {
        viewBinding.btnSendMessage.setEnabled(s.toString().trim().length() > 0);
    }
}

File: app/src/main/java/com/riders/thelab/ui/kat/KatActivity.kt
package com.riders.thelab.ui.kat

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.database.*
import com.riders.thelab.core.utils.LabDeviceManager
import com.riders.thelab.core.utils.UIManager
import com.riders.thelab.data.remote.dto.kat.Kat
import com.riders.thelab.databinding.ActivityKatBinding
import dagger.hilt.android.AndroidEntryPoint
import timber.log.Timber
import java.util.*

@AndroidEntryPoint
class KatActivity : AppCompatActivity(), TextWatcher {

    companion object {
        var SENDER_ID: String =
            LabDeviceManager.getBrand().toString() + ", " + LabDeviceManager.getModel()
    }

    private lateinit var viewBinding: ActivityKatBinding

    private var fetchedKatMessageList: List<Kat>? = null
    private var currentMessageList: List<Kat>? = null
    private var mAdapter: KatMessagesAdapter? = null

    private var database: FirebaseDatabase? = null
    private var messageRef: DatabaseReference? = null

    private var katModel: Kat? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = ActivityKatBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)

        viewBinding.btnSendMessage.isEnabled = false

        initMessageRecyclerView()

        // Write a message to the database
        database = FirebaseDatabase.getInstance()
        messageRef = database!!.getReference("messages")
        Timber.d("setupView()")
        val messagesQuery: Query = messageRef!!.ref

        // Build object
        initKatObject()


        // Read from the database
        messageRef!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                Timber.d("ValueEventListener - onDataChange()")

                // This method is called once with the initial value and again
                // whenever data at this location is updated.
                val genericTypeIndicator: GenericTypeIndicator<HashMap<String, Any>> =
                    object : GenericTypeIndicator<HashMap<String, Any>>() {}
                val value = dataSnapshot.getValue(genericTypeIndicator)

                // Get keys and values
                if (null != value) {
                    fetchedKatMessageList = ArrayList()
                    fetchedKatMessageList = Kat.buildKatMessagesList(value)
                    updateUI()
                } else {
                    removeAll()
                }
            }

            override fun onCancelled(error: DatabaseError) {
                // Failed to read value
                Timber.e(error.toException(), "Failed to read value.")
            }
        })
        viewBinding.etMessage.addTextChangedListener(this)


        // Handle send message button event
        viewBinding.btnSendMessage.setOnClickListener {
            val messageContent =
                if (Objects.requireNonNull(viewBinding.etMessage.text).toString().trim { it <= ' ' }
                        .isNotEmpty())
                    viewBinding.etMessage.text.toString().trim()
                else ""

            clearEditTextField()

            UIManager.hideKeyboard(this@KatActivity, findViewById(android.R.id.content))

            // Add complement values to kat object
            katModel!!.message = messageContent
            katModel!!.timestamp = System.currentTimeMillis()
            val key = messageRef!!.push().key
            katModel!!.messageId = key
            messageRef!!
                .child(key!!)
                .setValue(katModel)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
            }
        }
        return true
    }

    private fun initMessageRecyclerView() {
        Timber.d("initMessageRecyclerView()")
        currentMessageList = ArrayList()
        mAdapter = KatMessagesAdapter(this, currentMessageList as ArrayList<Kat>)
        val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        viewBinding.rvMessages.layoutManager = linearLayoutManager
        viewBinding.rvMessages.adapter = mAdapter
    }

    private fun initKatObject() {
        Timber.d("initKatObject()")
        katModel = Kat()
        katModel!!.senderId = SENDER_ID
        katModel!!.chatId = Random().nextInt(1000).toString()
    }

    private fun clearEditTextField() {
        Timber.d("clearEditTextField()")
        viewBinding.etMessage.setText("")
    }

    private fun updateUI() {
        Timber.d("updateUI()")
        Timber.d("isFromEventListener")
        // If it's first time, currentList is empty then populate list with fetched item
        if (currentMessageList!!.isEmpty()) {
            Timber.d("is first time")
            currentMessageList = fetchedKatMessageList
            currentMessageList?.let { mAdapter!!.populateAllItems(it) }
        } else {
            Timber.e("NOT first time")
            val kat = fetchedKatMessageList!![fetchedKatMessageList!!.size - 1]
            Timber.e("last item : %s", kat.toString())
            mAdapter!!.populateItem(kat)
            smoothScrollToLastItem()
        }
    }

    private fun smoothScrollToLastItem() {
        runOnUiThread {
            viewBinding.rvMessages.smoothScrollToPosition(
                Objects.requireNonNull(viewBinding.rvMessages.adapter).itemCount - 1
            )
        }
    }

    private fun compareListAndGetAddedItem(
        currentList: List<Kat>,
        fetchedList: List<Kat>
    ): List<Kat> {
        Timber.d("compareListAndGetAddedItem()")
        val differences: MutableList<Kat> = ArrayList(fetchedList)
        differences.removeAll(currentList)
        for (katElement in differences) {
            Timber.e("Differences : %s", katElement.toString())
        }
        return differences
    }

    private fun removeAll() {
        mAdapter!!.removeAll()
    }

    override fun onDestroy() {
        super.onDestroy()
        database = null
        messageRef = null
        fetchedKatMessageList = null
        currentMessageList = null
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        viewBinding.btnSendMessage.isEnabled = s.toString().trim { it <= ' ' }.isNotEmpty()
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        viewBinding.btnSendMessage.isEnabled = s.toString().trim { it <= ' ' }.isNotEmpty()
    }

    override fun afterTextChanged(s: Editable) {
        viewBinding.btnSendMessage.isEnabled = s.toString().trim { it <= ' ' }.isNotEmpty()
    }
}